### PR TITLE
osd: an fmtlib formatter for OpSchedulerItem

### PR DESCRIPTION
**Pull request by ronen-fr**
extracted, being a separate issue, from the scrub-scheduling PR

### CODE REVIEW

**Comment by ronen-fr** [comment]
extracted, being a separate issue, from the scrub-scheduling PR

**Review by rzarzynski — APPROVED** [review]

**Comment by ronen-fr** [comment]
jenkins test api

**Comment by ronen-fr** [comment]
jenkins test api

**Comment by ronen-fr** [comment]
jenkins test make check

**Review by athanatos — APPROVED** [review]

**Comment by ronen-fr** [comment]
jenkins test api

**Comment by ronen-fr** [comment]
Minimal Teuthology testing were performed - enough to guarantee that the code introduced compiles
and passes sanity checks. I will be merging based on that, as no logic changes are involved,

File: src/osd/scheduler/OpSchedulerItem.h [hunk 1]
```diff
@@ -601,7 +644,35 @@ class PGRecoveryMsg : public PGOpQueueable {
...
+    const auto pushes =
+	opsi.get_reserved_pushes()
+	    ? fmt::format(" reserved_pushes {}", opsi.get_reserved_pushes())
+	    : "";
+
+    return format_to(
+	ctx.out(), "OpSchedulerItem({} {} class_id {} prio {}{} cost {} e{}{})",
+	opsi.get_ordering_token(), opsi.qitem->print(),
+	static_cast<class_t>(opsi.get_scheduler_class()), opsi.get_priority(),
+	qos_cost, opsi.get_cost(), opsi.get_map_epoch(), pushes);
```
Comment (cbodley) [inline src/osd/scheduler/OpSchedulerItem.h:676]: i'm seeing compilation failures from this call on gcc 13.1.1. i think it needs to be qualified as `fmt::format_to()`. similar changes were made in https://github.com/ceph/ceph/pull/50438 to fix gcc13 build failures

```
/home/cbodley/ceph/src/osd/scheduler/OpSchedulerItem.h:672:21: error: call of overloaded ‘format_to(fmt::v9::basic_format_context<fmt::v9::appender, char>::it
erator, const char [59], const spg_t&, std::string, class_t, unsigned int, const std::__cxx11::basic_string<char>&, int, epoch_t, const std::__cxx11::basic_st
ring<char>&)’ is ambiguous
  672 |     return format_to(
      |            ~~~~~~~~~^
  673 |         ctx.out(), "OpSchedulerItem({} {} class_id {} prio {}{} cost {} e{}{})",
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  674 |         opsi.get_ordering_token(), opsi.qitem->print(),
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  675 |         static_cast<class_t>(opsi.get_scheduler_class()), opsi.get_priority(),
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  676 |         qos_cost, opsi.get_cost(), opsi.get_map_epoch(), pushes);
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/cbodley/ceph/src/fmt/include/fmt/core.h:3233:17: note: candidate: ‘OutputIt fmt::v9::format_to(OutputIt, format_string<T ...>, T&& ...) [with OutputIt =
 appender; T = {const spg_t&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned char, unsigned int, const std::__cxx11
::basic_string<char, std::char_traits<char>, std::allocator<char> >&, int, unsigned int, const std::__cxx11::basic_string<char, std::char_traits<char>, std::a
llocator<char> >&}; typename std::enable_if<detail::is_output_iterator<OutputIt, char>::value, int>::type <anonymous> = 0; format_string<T ...> = basic_format
_string<char, const spg_t&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned char, unsigned int, const std::__cxx11::
basic_string<char, std::char_traits<char>, std::allocator<char> >&, int, unsigned int, const std::__cxx11::basic_string<char, std::char_traits<char>, std::all
ocator<char> >&>]’
 3233 | FMT_INLINE auto format_to(OutputIt out, format_string<T...> fmt, T&&... args)
      |                 ^~~~~~~~~
/usr/include/c++/13/format:3745:5: note: candidate: ‘_Out std::format_to(_Out, format_string<_Args ...>, _Args&& ...) [with _Out = fmt::v9::appender; _Args = 
{const spg_t&, __cxx11::basic_string<char, char_traits<char>, allocator<char> >, unsigned char, unsigned int, const __cxx11::basic_string<char, char_traits<ch
ar>, allocator<char> >&, int, unsigned int, const __cxx11::basic_string<char, char_traits<char>, allocator<char> >&}; format_string<_Args ...> = basic_format_
string<char, const spg_t&, __cxx11::basic_string<char, char_traits<char>, allocator<char> >, unsigned char, unsigned int, const __cxx11::basic_string<char, ch
ar_traits<char>, allocator<char> >&, int, unsigned int, const __cxx11::basic_string<char, char_traits<char>, allocator<char> >&>]’
 3745 |     format_to(_Out __out, format_string<_Args...> __fmt, _Args&&... __args)
      |     ^~~~~~~~~
```
i've opened https://tracker.ceph.com/issues/61975 to track this

Comment (cbodley) [reply src/osd/scheduler/OpSchedulerItem.h:676]: see https://github.com/ceph/ceph/pull/52417

Comment (ronen-fr) [reply src/osd/scheduler/OpSchedulerItem.h:676]: @cbodley - thanks for fixing that.

**Comment by cbodley** [comment]
@ronen-fr separately, the centos 9 shaman builds from https://shaman.ceph.com/builds/ceph/wip-rf-opitem-fmt/ had failed too:
```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/18.0.0-4794-g05efd9e5/rpm/el9/BUILD/ceph-18.0.0-4794-g05efd9e5/src/osd/scheduler/OpSchedulerItem.h:236:23:   required from here
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/18.0.0-4794-g05efd9e5/rpm/el9/BUILD/ceph-18.0.0-4794-g05efd9e5/src/osd/scheduler/OpSchedulerItem.h:236:23:   in ‘constexpr’ expansion of ‘fmt::v8::basic_format_string<char, const Message&>("PGOpItem(op={})")’
/usr/include/fmt/core.h:2673:12: error: use of deleted function ‘fmt::v8::detail::fallback_formatter<T, Char, Enable>::fallback_formatter() [with T = Message; Char = char; Enable = void]’
 2673 |   auto f = conditional_t<has_formatter<mapped_type, context>::value,
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2674 |                          formatter<mapped_type, char_type>,
      |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2675 |                          fallback_formatter<T, char_type>>();
      |                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/fmt/core.h:1041:3: note: declared here
 1041 |   fallback_formatter() = delete;
      |   ^~~~~~~~~~~~~~~~~~
```
https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos9,DIST=centos9,MACHINE_SIZE=gigantic/71252//consoleFull

@yuriw caught this trying to schedule tests against https://github.com/ceph/ceph/pull/50441. i opened https://tracker.ceph.com/issues/61977 for this one

**Comment by ronen-fr** [comment]
@cbodley - that's a bit stranger... did not appear in my builds. Will try to understand tomorrow morning.

**Pull request merged**